File: src/main/java/org/example/bankramenserver/global/error/GlobalExceptionHandler.java
package org.example.bankramenserver.global.error;

import jakarta.validation.ConstraintViolationException;
import lombok.extern.slf4j.Slf4j;
import org.example.bankramenserver.global.error.exception.GlobalException;
import org.example.bankramenserver.global.error.exception.ErrorCode;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
import org.springframework.web.bind.MissingServletRequestParameterException;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(GlobalException.class)
    public ResponseEntity<ErrorResponse> handleGlobalException(GlobalException e) {
        log.error("GlobalException: {}", e.getMessage());
        ErrorCode errorCode = e.getErrorCode();
        return ResponseEntity
                .status(errorCode.getStatus())
                .body(ErrorResponse.of(errorCode));
    }

    @ExceptionHandler({
            MethodArgumentNotValidException.class,
            ConstraintViolationException.class,
            HttpMessageNotReadableException.class,
            MethodArgumentTypeMismatchException.class,
            MissingServletRequestParameterException.class
    })
    public ResponseEntity<ErrorResponse> handleInvalidRequest(Exception e) {
        log.error("InvalidRequestException: {}", e.getMessage());
        return ResponseEntity
                .status(ErrorCode.INVALID_REQUEST.getStatus())
                .body(ErrorResponse.of(ErrorCode.INVALID_REQUEST));
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleException(Exception e) {
        log.error("UnhandledException: {}", e.getMessage());
        return ResponseEntity
                .status(ErrorCode.INTERNAL_SERVER_ERROR.getStatus())
                .body(ErrorResponse.of(ErrorCode.INTERNAL_SERVER_ERROR));
    }
}

File: src/main/kotlin/org/example/bankramenserver/global/error/GlobalExceptionHandler.kt
package org.example.bankramenserver.global.error

import jakarta.validation.ConstraintViolationException
import org.example.bankramenserver.global.error.exception.ErrorCode
import org.example.bankramenserver.global.error.exception.GlobalException
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.MissingServletRequestParameterException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException

@RestControllerAdvice
class GlobalExceptionHandler {
    private val log = LoggerFactory.getLogger(javaClass)
    @ExceptionHandler(GlobalException::class) fun handleGlobalException(e: GlobalException): ResponseEntity<ErrorResponse> { log.error("GlobalException: {}", e.message); return ResponseEntity.status(e.errorCode.status).body(ErrorResponse.of(e.errorCode)) }
    @ExceptionHandler(MethodArgumentNotValidException::class, ConstraintViolationException::class, HttpMessageNotReadableException::class, MethodArgumentTypeMismatchException::class, MissingServletRequestParameterException::class)
    fun handleInvalidRequest(e: Exception): ResponseEntity<ErrorResponse> { log.error("InvalidRequestException: {}", e.message); return ResponseEntity.status(ErrorCode.INVALID_REQUEST.status).body(ErrorResponse.of(ErrorCode.INVALID_REQUEST)) }
    @ExceptionHandler(Exception::class) fun handleException(e: Exception): ResponseEntity<ErrorResponse> { log.error("UnhandledException: {}", e.message); return ResponseEntity.status(ErrorCode.INTERNAL_SERVER_ERROR.status).body(ErrorResponse.of(ErrorCode.INTERNAL_SERVER_ERROR)) }
}
